src/imitation/data/types.py

```diff
@@ -21,8 +21,8 @@ class Trajectory:
     acts: np.ndarray
     """Actions, shape (trajectory_len, ) + action_shape."""
 
-    infos: Optional[Sequence[dict]]
-    """A sequence of info dicts, length trajectory_len."""
+    infos: Optional[np.ndarray]
```

Comment:
`Sequence[dict]` seems like a much more natural type than `np.ndarray`, although admittedly `np.ndarray` can store `dict` objects. What is the advantage of `ndarray`?

Reply:
There's logic in `EpochOrderDataset`, `Buffer`, and `rollout` that assumes ndarray. For example, `EpochOrderDataset` use numpy indexing to select a random sample: `self.data_map["infos"][perm_idx]`. `Buffer` saves its data in a ndarray, returns an ndarray, and will ask for the dtype of `infos` when allocating its underlying arrays used for ring buffers.

We could support `infos: Sequence[dict]` by adding edge cases in these files but it seems easier to just use `np.ndarray` like all the other fields in `Transitions`.

Reply:
My understanding was that `EpochOrderDataset` and `Buffer` both operate on `Dict[str, np.ndarray]` and we use `TransitionsDictDatasetAdaptor` to convert them to/from `types.Transitions`. So wouldn't the conversion code just need to lie in `TransitionsDictDatasetAdaptor`? Perhaps we think it's still not worth adding that code, but this feels quite different than needing to make changes in multiple classes.

What part of `rollout` depends on this? I think maybe `finish_trajectory` in `TrajectoryAccumulator` would need to be changed.

Reply:
I guess I'm OK with it being an `np.ndarray` since there don't seem to be any major downsides: see e.g. https://stackoverflow.com/questions/26768013/what-are-the-benefits-drawbacks-of-a-list-of-lists-compared-to-a-numpy-array-o https://stackoverflow.com/questions/26767694/what-are-the-benefits-drawbacks-of-a-list-of-lists-compared-to-a-numpy-array-o

But I'd still like to understand better what the problem with `Sequence[dict]` is here.

Reply:
Agree that `TransitionsDictDatasetAdaptor` can handle the `EpochOrderDataset` case.

Places to add info conversions should we stick with Sequence[dict]:
  * `ReplayBuffer` accepts `Transitions` directly. (So need to convert infos to `array(dtype=object)`.)
  *  rollout: `np.concat` and `np.stack` in `finish_trajectory` and `flatten_trajectories_*`.
 
I think I had a cached belief that conversion is more onerous than it actually is because I was thinking about doing this with `Optional[Sequence[dict]]` instead, so I'd have to convert from None to array of empty dicts in several places.

Reply:
Thanks for the examples. I think I'm OK with either option now so you can make the final call.